### PR TITLE
Fix the Thread Page to get all Threads from the Database and map

### PR DESCRIPTION
**Pull request by Harrylever**
## Thread Page Not Working Fixes
_Page is expected to_
- Display all threads associated with member_id of a particular organisation
- Collect and organize threads based on the last edited/last created
<br/>

Before
Screenshot 2022-11-26 09:43:57

After
Screenshot 2022-11-26 09:47:02

## Backend
First of all, for the organization we created after creating accounts, the Database was not having any Messages with threads, so we created a **PUT** end-point to create threads, but no place to implement the feature on the frontend given that It's also not our Task, we used ThunderClient to make a post to the End-point to create threads on messages manually like so
> Route Details 
1. PUT Threads End-Point
Linear Board Task: https://linear.app/team-powerdrill/issue/BAC-16/create-put-end-point-for-thread-messages
Route: `"/org/{org_id}/rooms/{room_id}/messages/{message_id}/edithreads`
Location: backend/endpoints/threads.py
Method: **`PUT`** 
Returns: JSON Object of Updated Threads
*Route Request Data:* Data Type/Schema: MessageRequest
```
{
  "sender_id":"63686b924746182adae96b8e",
  "timestamp":1669287132948,
  "emojis":[],
  "richUiData": {
    "blocks": [
      {
        "key":"c53ft",
        "text":"What said is not correct",
        "type":"unstyled",
        "depth":0,
        "inlineStyleRanges":[],
        "entityRanges":[],
        "data":{}
      }
    ],
  "entityMap":{}
  }
}
```
Extra Info: We created a function in utils.message_utils.py specifically for updating threads `update_message_threads(org_id: str, message_id: str, message: dict[str, Any]) -> dict[str, Any]:` but imported as `edit_message_threads` in threads.py
Linear Board Task: https://linear.app/team-powerdrill/issue/BAC-19/create-the-backend-shadow-function-to-edit-message-to-append-the
Image 
code

After creating the Post End-point, we would need to get all messages associated with the specific member_id of the current user in the organization with their Threads, not an empty array. So for that, we created a **GET** end-point and its co-side controller to achieve this<br/>
3. GET Threads End-Point
Linear Board Task: https://linear.app/team-powerdrill/issue/BAC-18/create-get-end-point-for-thread-messages
Route: `/org/{org_id}/member/{member_id}/threads`
Location: backend/endpoints/threads.py
Method: **`GET`** 
Returns: JSON Object of Messages with Threads _beginning a not-empty array_
*Route Request Data:* Data Type/Schema: org_id: str, member_id: str
Image
code2

## Frontend
We edited the threads/index.jsx for the threads page. We created component elements in the components folder, for mapping through the thread's data to display each of them with similar visualisations.
#### created
- frontend/zc_messaging/src/pages/threads/threads.css
- frontend/zc_messaging/src/component/ThreadList/index.jsx
- frontend/zc_messaging/src/component/ThreadList/threadList.css
- frontend/zc_messaging/src/component/ThreadList/ThreadItem/index.jsx
- frontend/zc_messaging/src/component/ThreadList/ThreadItem/threadItem.css
- frontend/zc_messaging/src/pages/threads/threads.utils.js
#### edited
- frontend/zc_messaging/src/pages/threads/index.jsx
- frontend/zc_messaging/src/pages/index.js
- frontend/zc_messaging/src/component/index.js

1. Create the Mapping Function For the Loaded Array of Threads between the Pages and Component: pages/threads/index.jsx - component/ThreadList/index.jsx - component/ThreadList/ThreadItem/index.jsx and style the components
Linear board Task: https://linear.app/team-powerdrill/issue/FRO-35/create-and-implement-design-to-display-the-thread-page-for-zuri-chat
Linear Board Task: https://linear.app/team-powerdrill/issue/FRO-36/create-the-mapping-function-for-the-loaded-array-of-threads-between

2. Create the function to fetch the Threads Data from the Backend using Axios and modify the created_at to display properly using momentjs - Include functions to handle the loading of the Data while Page is loading. Create the Asynchronous function to fetch data in threads.utils.js in a separate file
Linear Board Task: https://linear.app/team-powerdrill/issue/FRO-37/create-the-function-to-fetch-the-threads-data-from-the-backend-using
Linear Board Task: https://linear.app/team-powerdrill/issue/FRO-38/create-the-asynchronous-function-to-fetch-data-in-threadsutilsjs

3. Create an elaborate Pull Request Description
Linear Board Task: https://linear.app/team-powerdrill/issue/FRO-39/create-an-elaborate-pull-request-description

## Note:
The zc_messaging plugin imported from  packages/root-config/src/assets/plugin.importmap.json was changed to from 
`"@zuri/zuri-plugin-messaging": "https://chat.zuri.chat/zuri-messaging.js"` to  `"@zuri/zuri-plugin-messaging": "http://localhost:5013/zuri-messaging.js"` for the changes in zc_messaging running on our local machine to be passed forward

### CODE REVIEW

File: frontend/.husky/pre-commit [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 echo '🏗️👷 Linting, testing and building your project before committing'
 
-cd frontend && yarn lint-staged
+# cd frontend && yarn lint-staged
```
Comment (ickynavigator) [inline frontend/.husky/pre-commit]: why?

File: frontend/package.json [hunk 2]
```diff
@@ -47,7 +47,8 @@
     "yarn-run-all": "^3.1.1"
   },
   "dependencies": {
-    "eslint": "^8.0.1"
+    "eslint": "^8.0.1",
+    "moment": "^2.29.4"
```
Comment (ickynavigator) [inline frontend/package.json]: this is not the right place to install that package

File: frontend/zc_messaging/src/component/ThreadList/ThreadItem/index.jsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import React from "react"
+
```
Comment (ickynavigator) [inline frontend/zc_messaging/src/component/ThreadList/ThreadItem/index.jsx]: why the spacing between imports?

File: frontend/zc_messaging/src/component/ThreadList/index.jsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import React from "react"
+import ThreadItem from "./ThreadItem"
+import "./threadList.css"
+
+const ThreadList = ({ threadListData }) => (
+  <div className="threadList-Wrap">
+    {threadListData.map((thread, index) => (
+      <ThreadItem key={index} Thread={thread} />
```
Comment (ickynavigator) [inline frontend/zc_messaging/src/component/ThreadList/index.jsx]: an index is not a key

File: frontend/zc_messaging/src/pages/threads/index.jsx [hunk 5]
```diff
@@ -1,13 +1,65 @@
...
+import "./threads.css"
+import { ThreadList } from "../../component"
+
 import generatePageTitle from "../../utils/generatePageTitle"
+import { getThreadHandler } from "./threads.utils"
 
 const Threads = () => {
+  const authUser = useSelector(state => state.authUser)
+  const [threads, setThreads] = useState([])
+  const jasonaParseData = data => {
```
Comment (ickynavigator) [inline frontend/zc_messaging/src/pages/threads/index.jsx]: jasona? 😂

File: frontend/zc_messaging/src/pages/threads/index.jsx [hunk 6]
```diff
@@ -1,13 +1,65 @@
...
+        let dumpy_threads_Arr = []
+        for (let i = 0; i < threads_array.length; i++) {
+          for (let j = 0; j < threads_array[i].threads.length; j++) {
+            dumpy_threads_Arr.push(threads_array[i].threads[j])
+          }
+        }
+        setThreads(dumpy_threads_Arr)
+      })
+    } else {
+      console.warn("Nothing was returned")
```
Comment (ickynavigator) [inline frontend/zc_messaging/src/pages/threads/index.jsx]: why is this a warning?

File: frontend/zc_messaging/src/pages/threads/index.jsx [hunk 7]
```diff
@@ -1,13 +1,65 @@
...
+    }
+  }, [])
+
+  const getThreadFunc = (orgId, senderId) => {
+    console.warn(authUser)
```
Comment (ickynavigator) [inline frontend/zc_messaging/src/pages/threads/index.jsx]: same here

File: frontend/zc_messaging/src/pages/threads/threads.utils.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import axios from "axios"
+import { BASE_URL } from "../../utils/constants"
+
+/**
```
Comment (ickynavigator) [inline frontend/zc_messaging/src/pages/threads/threads.utils.js]: Great use of js docs

File: frontend/zc_messaging/src/pages/threads/threads.utils.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+const getThreadHandler = async (orgId, senderId) => {
+  try {
+    if (orgId && senderId) {
+      const getThreadRequest = await axios.get(
+        `http://localhost:8000/api/v1/org/${orgId}/member/${senderId}/threads`
+      )
+      return getThreadRequest.data
+    }
+    throw new Error("Invalid arguments")
```
Comment (ickynavigator) [inline frontend/zc_messaging/src/pages/threads/threads.utils.js]: handle the error with a toast

**Review by ickynavigator — CHANGES_REQUESTED** [review]

File: frontend/zc_messaging/src/pages/threads/index.jsx (hunk 5 above)
Comment (JulianaSau) [reply frontend/zc_messaging/src/pages/threads/index.jsx]: 🤣🤣🤣

File: frontend/zc_messaging/src/pages/threads/threads.utils.js (hunk 8 above)
Comment (Harrylever) [reply frontend/zc_messaging/src/pages/threads/threads.utils.js]: Thank you Sir

File: frontend/zc_messaging/src/pages/threads/threads.utils.js (hunk 9 above)
Comment (Harrylever) [reply frontend/zc_messaging/src/pages/threads/threads.utils.js]: Okay Sir

File: frontend/zc_messaging/src/pages/threads/index.jsx (hunk 7 above)
Comment (Harrylever) [reply frontend/zc_messaging/src/pages/threads/index.jsx]: I'll take it out. console.log was disabled on eslint that's why I was using .warn to read out errors.

File: frontend/.husky/pre-commit (hunk 1 above)
Comment (Harrylever) [reply frontend/.husky/pre-commit]: Was blocking my push and I was a little impatient. but I've uncommented it. I'll run linter on my code to make sure the pre-commit works

File: frontend/package.json (hunk 2 above)
Comment (Harrylever) [reply frontend/package.json]: Moved to zc_messaging

File: frontend/zc_messaging/src/component/ThreadList/index.jsx (hunk 4 above)
Comment (Harrylever) [reply frontend/zc_messaging/src/component/ThreadList/index.jsx]: Changed to the unique id of each thread message

File: frontend/zc_messaging/src/component/ThreadList/ThreadItem/index.jsx (hunk 3 above)
Comment (Harrylever) [reply frontend/zc_messaging/src/component/ThreadList/ThreadItem/index.jsx]: Cleaned up.

**Pull request closed without merging**